runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +423,39 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range?
+--- @param on_parse fun(trees?: table<integer, TSTree>, err?: string)
+function LanguageTree:_async_parse(range, on_parse)
+  local ct = self:_buf().changedtick
+  local total_parse_time = 0
+  local redrawtime = vim.o.redrawtime
+
+  local function step()
+    -- If buffer was changed in the middle of parsing, reset parse state
+    if self:_buf().changedtick ~= ct then
+      self._parser:reset()
```

Comment:
This line causes the crash. Parser reset frees up the old subtree, and then it is used by query cursor somehow.

Reply:
![image](https://github.com/user-attachments/assets/1d67d761-db29-4bfe-a3cd-8515105e6068)

Reply:
My hypothesis is that previous highlight state with old query cursor is used. Because they are cleared asynchronously.

Reply:
@ribru17

Reply:
Also, does this mean anyone can trigger this from lua by using a stale `iter_captures()` iterator? 

All APIs are public IIRC

Reply:
> My hypothesis is that:
> 
> 1. Redraw starts (`_highlight_states` contain query cursor from previous redraw)
> 2. `Languagetree` resets the current parser, it frees the old tree (used by query cursor from previous redraw)
> 3. Parse times out, callback is never called. `_highlight_states` aren't cleared.
> 4. `on_line()` tries to use cursor from previous redraw. Cursor tries to read from the old tree. It is freed

Nice, yeah I believe this is exactly the issue

> Also, does this mean anyone can trigger this from lua by using a stale `iter_captures()` iterator?

I worried the same thing, and I believe it is possible but very unlikely, since most of the time you parse right before getting the root node, and then pass that node into the iterator. But that said this *really* should be fixed on the TS side of things, and I am surprised other editors have not raised this issue yet

Is there a way we can guard against this in neovim, maybe with `TSNode:has_changes()`?

Reply:
Keep changetick for a parser and check each call to the iterator that it is the same.

Reply:
@lewis6991

Does everything that depend on the tree (e.g. nodes and iterators) get invalidated after a reparse? How do the nodes/iterators check if the subtree they are referencing hasn't been freed? 

I tried changing the entire buffer buffer, reparsing, and then querying e.g. a child from an old node or the next capture from an old iterator. But they just return empty nodes and don't crash. 

It ssems strange that it only crashes when explicitly resetting the parser.

Reply:
`TSTree`s are reference counted (retain & release are +1 and -1 to refcount), so a more detailed :

1. First full parse creates a `tree1` [refcount=1], this tree is moved into Lua `_trees`

2. Redraw occurs. Since the parse succedes synchronously, highlight states are prepaed. A new query cursor is created (references the nodes in `tree1`, doesn't retain it, so refcount=1)
3. An edit occurs. `ts_tree_edit()` is called on `tree1`, but since its refcount=1, it is edited in place.
5. Redraw occurs and starts an async parse. Since the `changetick` at this point is already changed, the parser is not reset. A new parse starts, parser retains `tree1` [refcount=2]

7. Another edit occurs. `ts_tree_edit()` is called on `tree1`, and now since it has refcount>1, a new `tree2` [refcount=1] is created with modifications. `tree1` is released [refcount=1]
8. Async parse step is executed. Since the changetick is different, the parser is reset. Resetting the parser releases `tree1` [refcount=0]. `tree1` is freed.
9. Redraw occurs. Starts a new async parse but continues to redraw. Highlight states still keep a reference to query cursor with `tree1`. Query cursor tries to iterate nodes of `tree1` and crashes, because it's been freed.

This doesn't happen when parse is synchronous, since the tree edited in `ts_tree_edit()` has refcount=1 (because not in the parser) and thus modified in-place. And when it is freed by GC, it's children are not.

I.e. retaining the subtree before edit, and releasing after edit on a synchronous parse will also crash under the same circumstances:

<details>

<summary>Reproduction on (modified) master branch</summary>

```console
NVIM v0.11.0-dev-1417+g487c48ec86-dirty
Build type: Debug
LuaJIT 2.1.1734355927
```

```diff
diff --git a/src/nvim/lua/treesitter.c b/src/nvim/lua/treesitter.c
index 28ad2cf4d3..156a331ab4 100644
--- a/src/nvim/lua/treesitter.c
+++ b/src/nvim/lua/treesitter.c
@@ -746,6 +746,29 @@ static int tree_copy(lua_State *L)
   return 1;
 }
 
+
+typedef union {
+  uint64_t data;
+  const void *ptr;
+} Subtree;
+
+#define Array(T)       \
+  struct {             \
+    T *contents;       \
+    uint32_t size;     \
+    uint32_t capacity; \
+  }
+
+typedef Array(Subtree) SubtreeArray;
+typedef Array(Subtree) MutableSubtreeArray;
+
+typedef struct {
+  MutableSubtreeArray free_trees;
+  MutableSubtreeArray tree_stack;
+} SubtreePool;
+
+SubtreePool ts_subtree_pool_new(int);
+
 static int tree_edit(lua_State *L)
 {
   if (lua_gettop(L) < 10) {
@@ -765,7 +788,12 @@ static int tree_edit(lua_State *L)
   TSInputEdit edit = { start_byte, old_end_byte, new_end_byte,
                        start_point, old_end_point, new_end_point };
 
+  uint64_t subtree = *(uint64_t*)ud->tree;
+  ts_subtree_retain(subtree);
   ts_tree_edit(ud->tree, &edit);
+  SubtreePool pool = ts_subtree_pool_new(0);
+  ts_subtree_release(&pool, subtree);
+  ts_subtree_pool_delete(&pool);
 
   return 0;
 }
```

minimal.lua:
```lua
for name, url in pairs {
  nvim_treesitter = 'https://github.com/nvim-treesitter/nvim-treesitter.git'
} do
  local install_path = vim.fn.fnamemodify('nvim_issue/' .. name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.opt.runtimepath:append(install_path)
end

require'nvim-treesitter.configs'.setup {
  ensure_installed = { "lua" },
  highlight = { enable = true },
}

-----

local line = 'printf("some text");'
local lines = {}
for i = 1, 200 do lines[i] = line end
vim.api.nvim_buf_set_lines(0, 0, -1, true, lines)

vim.treesitter.start(0, 'c')
local parser = vim.treesitter.get_parser(0)
local tree = parser:parse(true)[1]
local q = vim.treesitter.query.parse('c', "((identifier) @cap)")
local iter = q:iter_captures(tree:root(), 0)
iter()

for depth = 0, 3 do
  vim.api.nvim_buf_set_text(0, 25 + depth, 4, 25 + depth, 5, {})
  parser:parse(true)
end

vim.keymap.set('n', 'a', function()
  for _ = 1, 50 do iter() end
end)
```

```shell
nvim --clean -u minimal.lua
```

Pressing 'a' will crash neovim

</details>

Reply:
Also everything that needs the tree should either retain it and release on GC (tree-sitter doesn't seem to have API for that). Or they need to be invalidated on tree changes & parses.

Reply:
Perhaps, but the userdata objects from treesitter are supposed to be low level, and the languagetree object is higher level which always invalidates on changes. Users of the userdata objects should be aware of this and make sure they don't hang on to references too long.

Reply:
Is it acceptable to move the `:reset()` call to invalidate since it is called by `_on_bytes()`? So that child parsers will be recursively reset as well per https://github.com/neovim/neovim/pull/31631#discussion_r1897850800

Reply:
@vanaigr I think this patch fixes the segfault in all cases

```patch
diff --git a/runtime/lua/vim/treesitter/query.lua b/runtime/lua/vim/treesitter/query.lua
index dbe3d54c2f..cf8620770b 100644
--- a/runtime/lua/vim/treesitter/query.lua
+++ b/runtime/lua/vim/treesitter/query.lua
@@ -893,6 +893,10 @@ function Query:iter_captures(node, source, start, stop)
   local match_preds = memoize(match_id_hash, self.match_preds, true)
 
   local function iter(end_line)
+    if node:has_changes() then
+      return
+    end
+
     local capture, captured_node, match = cursor:next_capture()
 
     if not capture then
@@ -965,6 +969,10 @@ function Query:iter_matches(node, source, start, stop, opts)
   local cursor = vim._create_ts_querycursor(node, self.query, start, stop, opts)
 
   local function iter()
+    if node:has_changes() then
+      return
+    end
+
     local match = cursor:next_match()
 
     if not match then
```

Reply:
Since `has_changes()` uses the subtree, it can also cause a crash: 
```lua
for name, url in pairs {
  nvim_treesitter = 'https://github.com/nvim-treesitter/nvim-treesitter.git'
} do
  local install_path = vim.fn.fnamemodify('nvim_issue/' .. name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.opt.runtimepath:append(install_path)
end

require'nvim-treesitter.configs'.setup {
  ensure_installed = { "lua" },
  highlight = { enable = true },
}

-----

local line = 'printf("some text");'
local lines = {}
for i = 1, 2000 do lines[i] = line end
vim.api.nvim_buf_set_lines(0, 0, -1, true, lines)

vim.treesitter.start(0, 'c')
local parser = vim.treesitter.get_parser(0)
local tree = parser:parse(true)[1]
local q = vim.treesitter.query.parse('c', "((identifier) @cap)")
local iter = q:iter_captures(tree:root(), 0)
local nodes = {}
local function addAll(node)
  table.insert(nodes, node:child())
  for child in node:iter_children() do
    addAll(child)
  end
end
addAll(tree:root())

for depth = 0, 30 do
  vim.api.nvim_buf_set_text(0, 25 + depth, 4, 25 + depth, 5, {})
  parser._parser:reset()
  parser._parse(parser, true, 3)
  collectgarbage("collect")
end

vim.keymap.set('n', 'a', function()
  for _, node in ipairs(nodes) do
    node:has_changes()
  end
end)
```

Reply:
Damn, how frustrating. I think we must wait until this is fixed in tree-sitter

Reply:
I think we can call `ts_tree_copy()` when we create a new lua node/querycursor, and store the tree clone internally. Copying a tree doesn't seem to be that expensive (atomic increment for subtree + copying the included ranges).

Reply:
Brilliant @vanaigr , thank you! I have fixed `iter_(captures|matches)`, let me know if you think the implementation should be adjusted.

Reply:
Note: other crashes occur even without async parsing, tracked by #31758

Reply:
> Brilliant @vanaigr , thank you! I have fixed `iter_(captures|matches)`, let me know if you think the implementation should be adjusted.

I think it's good :+1: